drivers/usb/udc/udc_dwc2.c

```diff
@@ -585,9 +585,21 @@ static void dwc2_handle_xfer_next(const struct device *dev,
 	if (USB_EP_DIR_IS_OUT(cfg->addr)) {
 		dwc2_prep_rx(dev, buf, cfg);
 	} else {
-		if (dwc2_tx_fifo_write(dev, cfg, buf)) {
-			LOG_ERR("Failed to start write to TX FIFO, ep 0x%02x",
-				cfg->addr);
+		int err = dwc2_tx_fifo_write(dev, cfg, buf);
+
+		if (err) {
+			if (err == -ENOENT) {
+				LOG_WRN("Cannot write to TX FIFO, ep 0x%02x inactive", cfg->addr);
+			} else {
+				LOG_ERR("Failed to start write to TX FIFO, ep 0x%02x (err: %d)",
+					cfg->addr, err);
+			}
+
+			if (udc_submit_ep_event(dev, udc_buf_get(dev, cfg->addr), -ECONNREFUSED)) {
+				LOG_ERR("Failed to submit EP event");
+			};
+			/* Early return to avoid setting endpoint as busy. */
+			return;
```

Comment:
"This is necessary to make application aware
of USB packets flow. Missing notification here results in e.g. lack of
sent callback for a HID input report provided by application to an USB
HID class instance."

This is not a notification of what is missing, but the removal of the transfer from the queue.

I suggest to write it a little differently:

```c
		if (err) {
			LOG_ERR("Failed to start write to TX FIFO, ep 0x%02x (err: %d)",
				cfg->addr, err)

			buf = udc_buf_get(dev, cfg->addr)
			if (udc_submit_ep_event(dev, buf, -ECONNREFUSED)) {
				LOG_ERR("Failed to submit endpoint event");
			};

			return;

```

Reply:
> I suggest to write it a little differently:

Added `buf` assignment. Please let's keep separate warning log for `-ENOENT` as it's not related to an actual error (I saw it if I disconnect USB cable while sending HID input reports over USB). I prefer not to emit error logs when there is no actual error (it's expected behavior here).

Reply:
> "This is necessary to make application aware of USB packets flow. Missing notification here results in e.g. lack of sent callback for a HID input report provided by application to an USB HID class instance."
> 
> This is not a notification of what is missing, but the removal of the transfer from the queue.

Depends on perspective. From our HID application's point of view, the missing sent callback for a submitted HID report was the problem (application was unable to track HID reports submitted to the USB next stack). Sill, since change is applied to UDC driver, your explanation may be a better fit here. Updated the commit message.

Reply:
I do not see it addressed. There is no reason for different LOGs, please change is as suggested.